dask/dataframe/dask_expr/_merge.py

```diff
@@ -441,25 +388,28 @@ def _lower(self):
                 right_index=right_index,
                 shuffle_left_on=shuffle_left_on,
                 shuffle_right_on=shuffle_right_on,
-                _npartitions=self.operand("_npartitions"),
+                _npartitions=shuffle_npartitions,
             )
-
-        if shuffle_left_on and not left_already_partitioned:
+        if shuffle_left_on and not (
+            left_already_partitioned and self.left.npartitions == shuffle_npartitions
+        ):
             # Shuffle left
             left = RearrangeByColumn(
                 left,
                 shuffle_left_on,
-                npartitions_out=self._npartitions,
+                npartitions_out=shuffle_npartitions,
                 method=shuffle_method,
                 index_shuffle=left_index,
             )
 
-        if shuffle_right_on and not right_already_partitioned:
+        if shuffle_right_on and not (
+            right_already_partitioned and self.right.npartitions == shuffle_npartitions
+        ):
```

Comment:
these shuffle already partitioned conditions are the only logic that changed slightly. I think everything else on this PR was just moving the various pieces of the base class _divisions to the subclasses this was actually relevant for, i.e. mostly BlockwiseMerge